great_expectations/rule_based_profiler/rule_based_profiler.py

```diff
@@ -907,8 +909,13 @@ def _apply_runtime_environment(
             domain_type_directives_list=domain_type_directives_list,
         )
 
-        rule: Rule
-        self._profiler_config.rules = {rule.name: rule.to_json_dict() for rule in rules}
+        try:
+            original_config_rules = self._profiler_config.rules
+            rule: Rule
+            self._profiler_config.rules = {rule.name: rule.to_json_dict() for rule in rules}
+            yield
```

Comment:
So this will return a generator with `None` in it? And this is required because we are using the `@contextmanager` decorator? And the decorator is required because we want to use `with` in our calls to `_apply_runtime_environment? And we want to use `with` because...you lost me there.

Reply:
This is a good question. I realize I should add an except close in case `to_json_dict` raises.

The purpose of a context manager is to wrap a block of code so one can do some stuff before and after that code executes. That is:
```
with my_context_manager:
    my python code
```
becomes 
``` 
<context manager code before yields>
my python code
<context manager code after yields>
```

Here I always want to make sure we reset `self._profiler_config.rules` to what it was before we changed it. I use the `try/finally` to make sure we reset this even when there is an exception. I should reraise that exception though.

In general, context managers also are able to return something. For example when you open a file you do `with open(file) as f:` and `f` is the file object. I don't care about this returning functionality here but because context managers can do this, they are a `generator iterator`. That lets them yield some value and then still do the cleanup part. Since I don't care about returning anything I just do a bare `yield` so the return type of the context manager is `Iterator[None]`. The iterator returned can only yield 1 thing with gets used in the `with ... as ...` assignment.

Note, all of this is a shortcut way to write a context manager using the `contextlib` package. The more conventional way to is to write class and define `__enter__` and `__exit__` methods:

```
class MyContextManager:
    def __enter__(self):
        python code to run before

    def __exit__(self, exc_type, exc_value, exc_tb):
        python code to run after
        exception handling also happens here which is what the exc_* args are for.
```